### ISSUE TITLE
fix: UnimplementedError when running web app hosted on Firebase hosting

### ISSUE THREAD

**Issue by sarbogast** [Bdaya-Dev/oidc#129]
**Description**

I am trying to replace openidconnect with oidc in my Flutter app, and when I run it locally in my Chrome browser on localhost, everything works great. But when I run the exact same app from Firebase hosting, I get plenty of UnimplementedErrors.

**Steps To Reproduce**

I don't have a clear scenario or a simple app to replicate this issue, as again when I run it locally, it all works fine.
I have a class that monitors the user's session, combining oidc to log in to Auth0 and Firebase auth.
When I initialize the user session, I call the lazy constructor to load the discovery document from Auth0:

```
_oidc = OidcUserManager.lazy(
      discoveryDocumentUri: Uri.parse(
        'https://${environment.auth0.domain}/.well-known/openid-configuration',
      ),
      clientCredentials: OidcClientAuthentication.none(
        clientId: environment.auth0.clientId,
      ),
      store: OidcDefaultStore(),
      settings: OidcUserManagerSettings(
        redirectUri:
            kIsWeb ? htmlPageLink : Uri.parse(environment.auth0.redirectUri),
        postLogoutRedirectUri: postLogoutRedirectUri,
        frontChannelLogoutUri: frontChannelLogoutUri,
        supportOfflineAuth: true,
        strictJwtVerification: true,
        sessionManagementSettings: OidcSessionManagementSettings(
          enabled: true,
        ),
        options: OidcPlatformSpecificOptions(
          web: OidcPlatformSpecificOptions_Web(
            navigationMode:
                OidcPlatformSpecificOptions_Web_NavigationMode.newPage,
          ),
        ),
        scope: scopes,
      ),
    );
```

Then I try to figure out if I already have some cached tokens to recover from by calling:

```
await _oidc.init();
      final storedIdToken = _oidc.currentUser?.idToken;
      final storedRefreshToken = _oidc.currentUser?.token.refreshToken;
...
```

But then inside of init and only on my Firebase hosting setup, I get the following exception:

```
UnimplementedError
wrapException@https://dev.myderbigum.com/main.dart.js?v=453:7392:43
listenToFrontChannelLogoutRequests$2@https://dev.myderbigum.com/main.dart.js?v=453:412497:15
OidcFlutter_listenToFrontChannelLogoutRequests@https://dev.myderbigum.com/main.dart.js?v=453:111210:46
listenToFrontChannelLogoutRequests$2@https://dev.myderbigum.com/main.dart.js?v=453:409646:16
init$0/$async$init$0<@https://dev.myderbigum.com/main.dart.js?v=453:411378:56
_wrapJsFunctionForAsync/$protected@https://dev.myderbigum.com/main.dart.js?v=453:10396:17
call$2@https://dev.myderbigum.com/main.dart.js?v=453:131663:12
...
```

I have tried to bypass this call to `listenToFrontChannelLogoutRequests` because I'm not sure I need it, so I set `frontChannelLogoutUri` to null when creating the instance of OidcUserManager. But then I get an UnimplementedError when calling `loginAuthorizationCodeFlow`:

```
UnimplementedError�[0m [js_primitives.dart:28:4](org-dartlang-sdk:///dart-sdk/lib/_internal/js_runtime/lib/js_primitives.dart)
�[38;5;196m├┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄�[0m [js_primitives.dart:28:4](org-dartlang-sdk:///dart-sdk/lib/_internal/js_runtime/lib/js_primitives.dart)
�[38;5;196m│ #0   wrapException@https://dev.myderbigum.com/main.dart.js?v=455:7392:43�[0m [js_primitives.dart:28:4](org-dartlang-sdk:///dart-sdk/lib/_internal/js_runtime/lib/js_primitives.dart)
�[38;5;196m│ #1   prepareForRedirectFlow$1@https://dev.myderbigum.com/main.dart.js?v=455:412530:15�[0m [js_primitives.dart:28:4](org-dartlang-sdk:///dart-sdk/lib/_internal/js_runtime/lib/js_primitives.dart)
�[38;5;196m│ #2   OidcFlutter_prepareForRedirectFlow@https://dev.myderbigum.com/main.dart.js?v=455:111207:46�[0m [js_primitives.dart:28:4](org-dartlang-sdk:///dart-sdk/lib/_internal/js_runtime/lib/js_primitives.dart)
�[38;5;196m│ #3   prepareForRedirectFlow$1@https://dev.myderbigum.com/main.dart.js?v=455:409679:16�[0m [js_primitives.dart:28:4](org-dartlang-sdk:///dart-sdk/lib/_internal/js_runtime/lib/js_primitives.dart)
�[38;5;196m│ #4   loginAuthorizationCodeFlow$body$OidcUserManagerBase/$async$loginAuthorizationCodeFlow$4$extraParameters$options$promptOverride$uiLocalesOverride<@https://dev.myderbigum.com/main.dart.js?v=455:410218:27�[0m [js_primitives.dart:28:4](org-dartlang-sdk:///dart-sdk/lib/_internal/js_runtime/lib/js_primitives.dart)
�[38;5;196m│ #5   _wrapJsFunctionForAsync/$protected@https://dev.myderbigum.com/main.dart.js?v=455:10396:17�[0m [js_primitives.dart:28:4](org-dartlang-sdk:///dart-sdk/lib/_internal/js_runtime/lib/js_primitives.dart)
�[38;5;196m│ #6   call$2@https://dev.myderbigum.com/main.dart.js?v=455:131663:12�[0m [js_primitives.dart:28:4](org-dartlang-sdk:///dart-sdk/lib/_internal/js_runtime/lib/js_primitives.dart)
�[38;5;196m│ #7   _asyncStartSync@https://dev.myderbigum.com/main.dart.js?v=455:10360:20�[0m
```

I suspect those two might be linked and maybe I'm doing something in the wrong order. But I'm not sure.

**Comment by ahmednfwela** [comment]
Hi @sarbogast 
What I am noticing here is that any calls to the underlying flutter platform are failing, specifically

https://github.com/Bdaya-Dev/oidc/blob/56433446586468e9d3d75937c6687b94dce7909d/packages/oidc_platform_interface/lib/src/platform.dart#L74-L118

this can only happen if the flutter platform didn't get initialized correctly, which is done on every platform package, e.g. web: 
https://github.com/Bdaya-Dev/oidc/blob/56433446586468e9d3d75937c6687b94dce7909d/packages/oidc_web/lib/oidc_web.dart#L10-L13


but I am not sure why this would ever happen, unless the oidc_web package somehow didn't get shipped as part of the `oidc` package? can you show me your `pubspec.yaml` ?

**Comment by sarbogast** [comment]
I thought it might have been because I was including oidc_web_core explicitly in my pubspec, but even after I removed it, I still get the same issue. Here are the dependencies in my pubspec.yaml:

```yaml
dependencies:
  flutter:
    sdk: flutter
  flutter_localizations:
    sdk: flutter
  flutter_web_plugins:
    sdk: flutter
  intl: ^0.19.0

  cupertino_icons: ^1.0.8
  logger: ^2.4.0
  flutter_bugfender: ^3.0.0+1
  provider: ^6.1.2
  responsive_builder: ^0.7.1
  package_info_plus: ^8.1.0
  flutter_svg: ^2.0.10+1
  smooth_page_indicator: ^1.2.0+3
  device_info_plus: ^11.1.0
  flutter_mailer: ^2.1.2
  multi_split_view: ^3.6.0
  extended_image: ^9.0.6
  flutter_form_builder: ^9.5.0
  form_builder_extra_fields: ^10.1.0 # complex dependency lock between form_build_extra_fields, flutter_typeahead and pointer_interceptor
    #git:
      #url: https://github.com/flutter-form-builder-ecosystem/form_builder_extra_fields.git
  flutter_typeahead: ^4.8.0 # complex dependency lock between form_build_extra_fields, flutter_typeahead and pointer_interceptor
  form_builder_validators: ^11.0.0
  form_builder_file_picker:
    git:
      url: https://github.com/DERBIGUM/form_builder_file_picker.git
  place_picker: #^0.9.20-nullsafety #not web-compatible
    git:
      url: https://github.com/DERBIGUM/locationpicker.git
      ref: nullsafety
  map_picker_flutter: #^1.1.0+1 #web-compatible and used as a web alternative to place_picker
    git:
      url: https://github.com/DERBIGUM/map_picker_flutter.git
  intl_phone_number_input: ^0.7.4
  phone_numbers_parser: ^9.0.1

  flutter_secure_storage: ^9.2.2
  jwt_decoder: ^2.0.1
  dart_jsonwebtoken: ^2.14.1
  go_router: 6.0.6 #newer versions have this bug: https://github.com/flutter/flutter/issues/112798
  url_launcher: ^6.3.1
  json_annotation: ^4.9.0
  universal_html: ^2.2.4
  uuid: ^4.5.1
  flinq: ^2.0.2
  latlong:
    git:
      url: https://github.com/sarbogast/dart-latlong.git
  image_picker: ^1.1.2
  photo_view: ^0.15.0
  share_plus: ^10.1.1
  cross_file: ^0.3.4+2
  fluttertoast: ^8.2.8
  retry: ^3.1.2
  pdf: ^3.11.1
  printing: ^5.13.4
  google_maps_flutter: ^2.9.0
  google_maps_flutter_web: ^0.5.4+2
  circle_flags: ^5.0.0
  flutter_google_places_web: #^1.0.0
    #path: ../flutter_google_places_web
    git:
      url: https://github.com/DERBIGUM/flutter_google_places_web.git
      #ref: c9a9637dc961b9a83955965069e32093efb11ab2
  flutter_hud: #^0.3.0
    git:
      url: https://github.com/DERBIGUM/flutter_hud.git
  another_flushbar: ^1.12.30
  path_provider: ^2.1.5
  html: ^0.15.5
  get_time_ago: ^2.0.0
  flutter_widget_from_html_core: ^0.15.2
  fwfh_svg: ^0.8.3
  pointer_interceptor: ^0.9.3+4 # complex dependency lock between form_build_extra_fields, flutter_typeahead and pointer_interceptor
  responsive_grid: ^2.4.4
  qr_flutter:
    git:
      url: https://github.com/lukef/qr.flutter.git
  video_player: ^2.9.2
  http: ^1.2.2
  country: ^4.1.9
  filesize_i18n:
    git:
      url: https://github.com/sarbogast/filesize_i18n.git
  font_awesome_flutter: ^10.8.0
  showcaseview: ^3.0.0
  shared_preferences: ^2.3.2
  badges: ^3.1.2
  flutter_slidable: ^3.1.1
  flutter_carousel_widget: ^3.1.0
  flutter_markdown: ^0.7.4+1
  reorderables: ^0.6.0
  google_places_autocomplete_text_field: ^0.1.3
  collection: ^1.18.0
  upgrader: ^11.3.0
  flutter_staggered_grid_view: ^0.7.0
  algolia_helper_flutter: ^1.1.0
  email_validator: ^3.0.0

  oidc: ^0.7.0
  oidc_default_store: ^0.2.0+9

  firebase_core: ^3.6.0
  cloud_firestore: ^5.4.4
  firebase_database: ^11.1.4
  firebase_auth: ^5.3.1
  cloud_functions: ^5.1.3
  firebase_analytics: ^11.3.3
  firebase_storage: ^12.3.4
  firebase_messaging: ^15.1.3
  #firebase_crashlytics: ^2.8.0
  firebase_in_app_messaging: ^0.8.0+8
  firebase_dynamic_links: ^6.0.8

dev_dependencies:
  flutter_lints: ^5.0.0
  json_serializable: ^6.8.0
  build_runner: ^2.4.13

dependency_overrides:
  web: ^1.1.0
  location: ^6.0.2
  flutter_keyboard_visibility: ^6.0.0
  intl: ^0.19.0
  uuid: ^4.5.1
  image: ^4.3.0
  http: ^1.2.2
  dio: ^5.7.0
  pdf_widget_wrapper: ^1.0.4
  flutter_colorpicker: ^1.1.0
  archive: ^3.6.1
  font_awesome_flutter:
    path: ../font_awesome_flutter
  showcaseview:
    git:
      url: https://github.com/chandrabezzo/flutter_showcaseview
      ref: bugfix/text-theme/removed-3.22
  photo_view:
    git:
      url: https://github.com/bluefireteam/photo_view
  circle_flags:
    #path: ../flutter_circle_flags
    git:
      url: https://github.com/DERBIGUM/flutter_circle_flags
  google_places_autocomplete_text_field:
    #path: ../google_places_autocomplete_text_field
    git:
      url: https://github.com/DERBIGUM/google_places_autocomplete_text_field.git
```

And in case it matters, here is the command with which I build my web version before deploying it to Firebase hosting:

```
flutter build web --source-maps --profile -t lib/main_dev.dart
```

**Comment by ahmednfwela** [comment]
to make sure, can you call this in your `main_dev.dart` ?

```dart
// add necessary imports
OidcPlatform.instance = OidcWeb();
```

if the error still persists after this, then it's a different problem

**Comment by sarbogast** [comment]
What is the import for OidcWeb()? Because I have tried importer oidc_web_core/oidc_web_core.dart but it's not defined there apparently.

**Comment by sarbogast** [comment]
Nevermind, I found it in oidc_web

**Comment by sarbogast** [comment]
I added the following to main_dev.dart:

```dart
if (kIsWeb) {
      OidcPlatform.instance = OidcWeb();
    } else if (Platform.isAndroid) {
      OidcPlatform.instance = OidcAndroid();
    } else if (Platform.isIOS) {
      OidcPlatform.instance = OidcIOS();
    } else {
      throw UnsupportedError('Unsupported platform');
    }
```

And I still have the same issue

**Comment by sarbogast** [comment]
Wait a second, I had another issue in my build, let me try again

**Comment by ahmednfwela** [comment]
can you also enable verbose logging like so:

https://github.com/Bdaya-Dev/oidc/blob/56433446586468e9d3d75937c6687b94dce7909d/packages/oidc/example/pubspec.yaml#L17C3-L17C10

https://github.com/Bdaya-Dev/oidc/blob/56433446586468e9d3d75937c6687b94dce7909d/packages/oidc/example/lib/main.dart#L16-L19

**Comment by sarbogast** [comment]
Could the issue be related to the fact that I had a lingering dependency on openidconnect in my pubspec? 

Because now that I removed it, it seems to work fine, even after I remove the piece of code to explicitly set OidcPlatform.instance and the dependencies on oidc_web, oidc_android and oidc_ios

**Comment by ahmednfwela** [comment]
That's great to hear, maybe there was a conflict somewhere that caused `oidc_web` to not get loaded, first time I have faced this problem to be honest.

feel free to close the issue if it's solved
